File: site/static/js/_ir_eval.js
"use strict";
/*
 * site/static/js/_ir_eval.js -- the ONE shared AST walker for the Toledo executable-equations
 * JavaScript twin (docs/EXECUTABLE_EQUATIONS_v0_1.md sec.5, stream S2).
 *
 * ES2020, zero dependencies, no `import`/`export`, no network fetch -- loaded by a bare
 * <script src="/static/js/_qfrac.js"> + <script src="/static/js/_ir_eval.js"> pair (this file
 * must load AFTER _qfrac.js, which defines the global `QFrac`/`QFracError` this file uses).
 * `node` runs this exact same file headlessly for the automated cross-check
 * (scripts/executable/crosscheck_runner.py, stream S3) -- a subprocess invocation of the
 * shipped browser file, never a second implementation for automation's sake (sec.5).
 *
 * Same node shape as scripts/executable/ir_kernel.py's `eval_node` (sec.3) -- this file is
 * that walker's JavaScript twin, not an independent redesign. No native single/double-
 * precision floating-point sin, cos, pi constant, sqrt, exp, or log built-in ever appears in
 * the evaluated/compared value path anywhere in this file -- every transcendental `fn` below
 * is computed via a genuinely
 * different algorithm FAMILY than ir_kernel.py's own (continued-fraction evaluation instead
 * of truncated power-series summation / Newton-Raphson), named in ALGORITHM_FAMILY_JS so a
 * build-time lint (scripts/executable/ir_eval.py::check_transcendental_lint) can assert the
 * two families are never the same string for the same `fn` (sec.3/sec.13).
 */
(function (root) {
  const QFrac = root.QFrac;
  const QFracError = root.QFracError;
  if (!QFrac) {
    throw new Error("_ir_eval.js requires _qfrac.js to be loaded first (defines global QFrac)");
  }

  class IRKernelError extends Error {}

  const ALLOWED_OPS = new Set(["add", "sub", "mul", "div", "pow", "neg", "const", "var", "call"]);
  const ALLOWED_FNS = new Set(["sqrt", "exp", "sin", "cos", "log", "pi_const"]);

  const DEFAULT_TERMS = 30;
  const WORKING_PRECISION_DIGITS = 60;

  // The algorithm family THIS file actually implements per transcendental `fn` -- a genuinely
  // different family than scripts/executable/ir_kernel.py::ALGORITHM_FAMILY_PY for every row,
  // by construction (continued-fraction evaluation vs. truncated power series / Newton).
  const ALGORITHM_FAMILY_JS = {
    pi_const: "arctan_continued_fraction",
    exp: "lambert_continued_fraction",
    sin: "tan_half_angle_continued_fraction",
    cos: "tan_half_angle_continued_fraction",
    sqrt: "generalized_continued_fraction",
    log: "continued_fraction_ln1p",
  };

  function require_(cond, msg) {
    if (!cond) throw new IRKernelError(msg);
  }

  function isqrtBigInt(n) {
    if (n < 0n) throw new IRKernelError("isqrt requires a nonnegative integer");
    if (n === 0n) return 0n;
    let x = n;
    let y = (x + 1n) / 2n;
    while (y < x) {
      x = y;
      y = (x + n / x) / 2n;
    }
    return x;
  }

  // ---------------------------------------------------------------------
  // Transcendental primitives -- continued-fraction family (ALGORITHM_FAMILY_JS).
  // Every intermediate value is rounded to WORKING_PRECISION_DIGITS via QFrac's own
  // roundToPrecision, the identical discipline ir_kernel.py applies on the Python side.
  // ---------------------------------------------------------------------

  /** Continued fraction for arctan(z): z / (1 + z^2*1^2/(3 + z^2*2^2/(5 + z^2*3^2/(7+...)))). */
  function arctanCF(z, terms) {
    let T = QFrac.fromInt(2 * terms + 1);
    for (let k = terms; k >= 1; k--) {
      const a = QFrac.fromInt(2 * (k - 1) + 1);
      const b = z.mul(z).mul(QFrac.fromInt(k * k));
      T = a.add(b.div(T)).roundToPrecision(WORKING_PRECISION_DIGITS);
    }
    return z.div(T).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  /** pi = 16*arctan(1/5) - 4*arctan(1/239) (Machin's identity), each arctan evaluated by
   * continued fraction instead of a power series -- a different algorithm family than
   * ir_kernel.py's own Machin-arctan-*series* reference. */
  function piConst(terms) {
    const a = arctanCF(QFrac.fromString("1/5"), terms);
    const b = arctanCF(QFrac.fromString("1/239"), terms);
    return QFrac.fromInt(16).mul(a).sub(QFrac.fromInt(4).mul(b)).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  /** Generalized continued fraction for sqrt(x): write x = p/q, n = p*q, a = isqrt(n),
   * r = n - a^2; sqrt(n) = a + r/(2a + r/(2a + ...)); sqrt(x) = sqrt(n)/q. A different
   * algorithm family than ir_kernel.py's own Newton-Raphson reference. */
  function sqrtFinite(x, terms) {
    require_(x.cmp(QFrac.fromInt(0)) >= 0, "sqrt domain requires a nonnegative rational");
    if (x.isZero()) return QFrac.fromInt(0);
    const p = x.num;
    const q = x.den;
    const n = p * q;
    const a = isqrtBigInt(n);
    const r = n - a * a;
    if (r === 0n) return new QFrac(a, q);
    let T = QFrac.fromInt(2n * a);
    const rFrac = new QFrac(r, 1n);
    for (let i = 0; i < terms; i++) {
      T = QFrac.fromInt(2n * a).add(rFrac.div(T)).roundToPrecision(WORKING_PRECISION_DIGITS);
    }
    const val = new QFrac(a, 1n).add(rFrac.div(T));
    return val.div(new QFrac(q, 1n)).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  /** Lambert's continued fraction for e^x:
   * e^x = 1 + 2x/((2-x) + x^2/(6 + x^2/(10 + x^2/(14 + ...)))). A different algorithm family
   * than ir_kernel.py's own halving-reduced Taylor series. */
  function expFinite(x, terms) {
    const x2 = x.mul(x);
    function denomAt(i) {
      return i === 0 ? QFrac.fromInt(2).sub(x) : QFrac.fromInt(4 * i + 2);
    }
    let T = denomAt(terms);
    for (let k = terms; k >= 1; k--) {
      T = denomAt(k - 1).add(x2.div(T)).roundToPrecision(WORKING_PRECISION_DIGITS);
    }
    return QFrac.fromInt(1).add(QFrac.fromInt(2).mul(x).div(T)).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  /** tan(x) via continued fraction: x/(1 - x^2/(3 - x^2/(5 - ...))); sin/cos recovered from
   * t = tan(x/2) via sin(x)=2t/(1+t^2), cos(x)=(1-t^2)/(1+t^2). A different algorithm family
   * than ir_kernel.py's own halving+double-angle Taylor series. */
  function tanCF(x, terms) {
    let T = QFrac.fromInt(2 * terms - 1);
    const x2 = x.mul(x);
    for (let k = terms - 1; k >= 1; k--) {
      T = QFrac.fromInt(2 * k - 1).sub(x2.div(T)).roundToPrecision(WORKING_PRECISION_DIGITS);
    }
    return x.div(T).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  function sinCosViaHalfAngle(x, terms) {
    const half = x.div(QFrac.fromInt(2));
    const t = tanCF(half, terms);
    const denom = QFrac.fromInt(1).add(t.mul(t));
    const s = QFrac.fromInt(2).mul(t).div(denom).roundToPrecision(WORKING_PRECISION_DIGITS);
    const c = QFrac.fromInt(1).sub(t.mul(t)).div(denom).roundToPrecision(WORKING_PRECISION_DIGITS);
    return [s, c];
  }

  function sinFinite(x, terms) {
    return sinCosViaHalfAngle(x, terms)[0];
  }

  function cosFinite(x, terms) {
    return sinCosViaHalfAngle(x, terms)[1];
  }

  /** ln(1+y) via continued fraction: y/(1 + 1^2*y/(2 + 1^2*y/(3 + 2^2*y/(4 + 2^2*y/(5+...))))).
   * x is first range-reduced into (1/2, 2] by halving/doubling against a once-computed ln(2)
   * (itself the same continued fraction at y=1) -- a power-of-2 scaling reduction, distinct
   * from ir_kernel.py's own repeated-sqrt reduction. A different algorithm family overall
   * than ir_kernel.py's own artanh-series-with-sqrt-reduction reference. */
  function ln1pCF(y, terms) {
    let D = QFrac.fromInt(terms);
    for (let k = terms - 1; k >= 1; k--) {
      const halfK = (k + 1) >> 1; // integer floor((k+1)/2) via bit shift -- no Math.* call
      const coeff = QFrac.fromInt(halfK ** 2);
      const N = coeff.mul(y);
      D = QFrac.fromInt(k).add(N.div(D)).roundToPrecision(WORKING_PRECISION_DIGITS);
    }
    return y.div(D).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  function logFinite(x, terms) {
    require_(x.cmp(QFrac.fromInt(0)) > 0, "log domain requires a positive rational");
    const ln2 = ln1pCF(QFrac.fromInt(1), terms); // ln(2)
    let reduced = x;
    let k = 0; // net factor is 2**k (reduced = x / 2**k)
    const two = QFrac.fromInt(2);
    const half = QFrac.fromString("1/2");
    let guard = 0;
    while (reduced.cmp(two) > 0 && guard < 4096) {
      reduced = reduced.div(two);
      k += 1;
      guard += 1;
    }
    while (reduced.cmp(half) < 0 && guard < 4096) {
      reduced = reduced.mul(two);
      k -= 1;
      guard += 1;
    }
    const y = reduced.sub(QFrac.fromInt(1));
    const lnReduced = ln1pCF(y, terms);
    return lnReduced.add(QFrac.fromInt(k).mul(ln2)).roundToPrecision(WORKING_PRECISION_DIGITS);
  }

  const TRANSCENDENTAL_FNS = {
    pi_const: (values, terms) => piConst(terms),
    sqrt: (values, terms) => sqrtFinite(values[0], terms),
    exp: (values, terms) => expFinite(values[0], terms),
    sin: (values, terms) => sinFinite(values[0], terms),
    cos: (values, terms) => cosFinite(values[0], terms),
    log: (values, terms) => logFinite(values[0], terms),
  };

  function callFn(fn, values, terms) {
    require_(ALLOWED_FNS.has(fn), `disallowed transcendental function: "${fn}"`);
    const expectedArity = fn === "pi_const" ? 0 : 1;
    require_(
      values.length === expectedArity,
      `"${fn}" takes exactly ${expectedArity} argument(s), got ${values.length}`
    );
    if (fn === "sqrt") {
      require_(values[0].cmp(QFrac.fromInt(0)) >= 0, "'sqrt' argument is outside its declared domain");
    }
    if (fn === "log") {
      require_(values[0].cmp(QFrac.fromInt(0)) > 0, "'log' argument is outside its declared domain");
    }
    return TRANSCENDENTAL_FNS[fn](values, terms);
  }

  // ---------------------------------------------------------------------
  // The generic AST walker -- the twin of ir_kernel.py::eval_node.
  // ---------------------------------------------------------------------

  function evalNode(node, env, terms) {
    if (typeof terms === "undefined") terms = DEFAULT_TERMS;
    require_(
      node && typeof node === "object" && typeof node.op === "string",
      "an IR node must be an object with an 'op' key"
    );
    const op = node.op;
    require_(ALLOWED_OPS.has(op), `disallowed IR operator: "${op}"`);

    if (op === "const") {
      return QFrac.fromString(node.value);
    }

    if (op === "var") {
      const name = node.name;
      require_(typeof name === "string" && Object.prototype.hasOwnProperty.call(env, name),
        `no bound value for variable "${name}"`);
      return env[name];
    }

    if (op === "call") {
      const fn = node.fn;
      const args = node.args || [];
      require_(Array.isArray(args), "a 'call' node's 'args' must be a list");
      const values = args.map((a) => evalNode(a, env, terms));
      return callFn(fn, values, terms);
    }

    const args = node.args;
    require_(Array.isArray(args) && args.length > 0, `"${op}" node requires a non-empty 'args' list`);
    const values = args.map((a) => evalNode(a, env, terms));

    if (op === "add") {
      let total = QFrac.fromInt(0);
      for (const v of values) total = total.add(v);
      return total;
    }
    if (op === "mul") {
      let product = QFrac.fromInt(1);
      for (const v of values) product = product.mul(v);
      return product;
    }
    if (op === "sub") {
      require_(values.length === 2, "'sub' requires exactly 2 args");
      return values[0].sub(values[1]);
    }
    if (op === "div") {
      require_(values.length === 2, "'div' requires exactly 2 args");
      require_(!values[1].isZero(), "division by zero");
      return values[0].div(values[1]);
    }
    if (op === "neg") {
      require_(values.length === 1, "'neg' requires exactly 1 arg");
      return values[0].neg();
    }
    if (op === "pow") {
      require_(values.length === 2, "'pow' requires exactly 2 args");
      const base = values[0];
      const exponent = values[1];
      require_(exponent.den === 1n, "'pow' exponent must evaluate to an integer rational");
      if (exponent.num < 0n) {
        require_(!base.isZero(), "cannot raise zero to a negative power");
      }
      return base.pow(exponent.num);
    }
    throw new IRKernelError(`unhandled operator: "${op}"`);
  }

  root.ToledoIREval = {
    IRKernelError,
    ALLOWED_OPS,
    ALLOWED_FNS,
    DEFAULT_TERMS,
    WORKING_PRECISION_DIGITS,
    ALGORITHM_FAMILY_JS,
    evalNode,
    piConst,
    sqrtFinite,
    expFinite,
    sinFinite,
    cosFinite,
    logFinite,
  };

  // Node/CommonJS export for headless invocation (scripts/executable/crosscheck_runner.py's
  // `node _ir_eval.js` subprocess call, and this stream's own node-run test suite) -- the
  // browser <script> tag path above (attaching to `root`/globalThis) is unaffected.
  if (typeof module !== "undefined" && module.exports) {
    module.exports = root.ToledoIREval;
  }
})(typeof globalThis !== "undefined" ? globalThis : this);

File: scripts/executable/ir_eval.py
#!/usr/bin/env python3
"""scripts/executable/ir_eval.py -- the ONE evaluator every consumer of a Toledo executable-
equations IR sidecar calls (docs/EXECUTABLE_EQUATIONS_v0_1.md sec.4, stream S2): the site
widget's build step, `toledo_eval` (mcp/toledo_mcp/server.py, stream S4), and
`scripts/executable/crosscheck_runner.py` (stream S3) all import this module rather than
re-implementing the walk -- never a second, parallel interpretation of the same IR (sec.13
item 4).

Loads one IR sidecar (registry/executable/<mangled-code>.json, sec.3 shape), validates it,
binds caller-supplied input values (accepted only as strings, parsed via
`ir_kernel.parse_rational` -- a JSON/Python float is rejected outright, never silently
coerced), walks the tree via `ir_kernel.eval_node`, and returns an exact-rational result plus
its `terms_used`/`error_bound` (transcendental case) or `None`/`None` (fully Q-exact case).

This module never edits registry/executable/*.json; it only reads a sidecar its caller
already loaded, or loads one itself from a path.
"""
from __future__ import annotations

import json
from fractions import Fraction
from pathlib import Path
from typing import Any, Mapping

try:
    from . import ir_kernel
except ImportError:
    # Loaded as a flat module rather than as part of the `scripts.executable` package --
    # e.g. via importlib.util.spec_from_file_location, mcp/toledo_mcp/core.py's own
    # established convention for reaching a repo script whose enclosing directories are not
    # guaranteed importable as regular packages (a real collision exists in at least one
    # environment on this workstation: a third-party `scripts` package installed in
    # site-packages shadows this repository's own top-level `scripts/` directory once any
    # sys.path entry ahead of the repo root supplies one). Fall back to a sibling-file import
    # so this module works identically either way.
    import sys as _sys
    from pathlib import Path as _Path

    _sys.path.insert(0, str(_Path(__file__).resolve().parent))
    import ir_kernel  # type: ignore[no-redef]

__all__ = [
    "IRValidationError",
    "REQUIRED_TOP_LEVEL_KEYS",
    "VALID_DOMAINS",
    "load_ir",
    "validate_ir_shape",
    "check_transcendental_lint",
    "evaluate",
    "evaluate_file",
]


class IRValidationError(ValueError):
    """The loaded IR sidecar (or the caller-supplied inputs) does not match the sec.3 schema
    this evaluator requires -- raised, never silently patched or guessed at."""


REQUIRED_TOP_LEVEL_KEYS = (
    "schema_version",
    "code",
    "root",
    "variables",
    "relation",
    "lhs",
    "rhs",
    "eligibility",
    "sample_inputs",
    "status",
)

VALID_DOMAINS = frozenset({"Z", "Z_pos", "Q", "Q_pos", "Q_nonzero"})

_VALID_STATUSES = frozenset({"candidate", "reviewed_eligible", "reviewed_rejected", "built"})


def _fail(message: str) -> None:
    raise IRValidationError(message)


def validate_ir_shape(ir: Mapping[str, Any]) -> None:
    """Structural validation against sec.3's IR schema. Does NOT check `status`/
    `eligibility.reviewed_by` gating for evaluation eligibility -- that fail-closed decision
    belongs to each caller (toledo_eval, crosscheck_runner), which know what status they are
    permitted to act on; this function only checks the document is well-formed."""
    if not isinstance(ir, Mapping):
        _fail("an IR sidecar must be a JSON object")

    missing = [k for k in REQUIRED_TOP_LEVEL_KEYS if k not in ir]
    if missing:
        _fail(f"IR sidecar missing required key(s): {missing}")

    if ir["schema_version"] != "executable-ir-0.1":
        _fail(f"unsupported schema_version: {ir['schema_version']!r}")

    if ir["status"] not in _VALID_STATUSES:
        _fail(f"unrecognised status: {ir['status']!r}")

    if ir["status"] in ("reviewed_eligible", "reviewed_rejected", "built"):
        eligibility = ir.get("eligibility") or {}
        if not eligibility.get("reviewed_by"):
            _fail(
                f"status {ir['status']!r} requires a non-empty eligibility.reviewed_by "
                "(sec.3: only a human registrar may write this status)"
            )
        if not eligibility.get("reviewed_at"):
            _fail(f"status {ir['status']!r} requires a non-null eligibility.reviewed_at")

    variables = ir["variables"]
    if not isinstance(variables, list) or not variables:
        _fail("'variables' must be a non-empty array")

    names_seen: set[str] = set()
    output_names: list[str] = []
    for v in variables:
        if not isinstance(v, Mapping) or "name" not in v or "domain" not in v or "role" not in v:
            _fail(f"malformed variable entry: {v!r}")
        name = v["name"]
        if name in names_seen:
            _fail(f"duplicate variable name: {name!r}")
        names_seen.add(name)
        if v["domain"] not in VALID_DOMAINS:
            _fail(f"variable {name!r} has an invalid domain: {v['domain']!r} (must be one of {sorted(VALID_DOMAINS)})")
        if v["role"] not in ("input", "output"):
            _fail(f"variable {name!r} has an invalid role: {v['role']!r}")
        if v["role"] == "output":
            output_names.append(name)

    if len(output_names) != 1:
        _fail(f"IR must declare exactly one output variable, found {len(output_names)}: {output_names}")

    if ir["relation"] != "eq":
        _fail(f"unsupported relation: {ir['relation']!r} (only 'eq' is implemented)")

    lhs = ir["lhs"]
    if lhs.get("op") != "var" or lhs.get("name") != output_names[0]:
        _fail(
            "'lhs' must be a bare {'op': 'var', 'name': <output variable>} reference -- "
            "this evaluator only computes the declared output from 'rhs', it does not solve "
            "an arbitrary equation for an unknown on either side"
        )

    check_transcendental_lint(ir)


def _collect_call_fns(node: Any, found: set[str]) -> None:
    if not isinstance(node, Mapping):
        return
    if node.get("op") == "call":
        fn = node.get("fn")
        if isinstance(fn, str):
            found.add(fn)
    for arg in node.get("args", []) or []:
        _collect_call_fns(arg, found)


def check_transcendental_lint(ir: Mapping[str, Any]) -> None:
    """Sec.3/sec.13's build-time lint: a 'call' node anywhere in lhs/rhs requires a non-null
    `transcendental` block naming exactly that function; `algorithm_py`/`algorithm_js` must
    match this kernel's own declared family (ir_kernel.ALGORITHM_FAMILY_PY/_JS) and must never
    name the same family as each other for the same fn. An equation mixing two distinct
    transcendental functions in one statement is out of scope for this narrow subset and is
    rejected here rather than silently under-checked."""
    fns_used: set[str] = set()
    _collect_call_fns(ir["lhs"], fns_used)
    _collect_call_fns(ir["rhs"], fns_used)

    transcendental = ir.get("transcendental")

    if not fns_used:
        if transcendental is not None:
            _fail("'transcendental' block present but no 'call' node exists in lhs/rhs")
        return

    if len(fns_used) > 1:
        _fail(
            f"more than one distinct transcendental function in one statement is out of scope "
            f"for this IR schema: {sorted(fns_used)}"
        )
    (fn,) = fns_used

    if transcendental is None:
        _fail(f"a 'call' node uses {fn!r} but no 'transcendental' block is present")

    for key in ("fn", "algorithm_py", "algorithm_js", "default_terms"):
        if key not in transcendental:
            _fail(f"'transcendental' block missing required key: {key!r}")

    if transcendental["fn"] != fn:
        _fail(f"'transcendental.fn' ({transcendental['fn']!r}) does not match the call node's fn ({fn!r})")

    if transcendental["algorithm_py"] == transcendental["algorithm_js"]:
        _fail(
            f"algorithm_py and algorithm_js name the same family ({transcendental['algorithm_py']!r}) "
            f"for {fn!r} -- a cross-check between one algorithm typed twice is not independent evidence"
        )

    expected_py = ir_kernel.ALGORITHM_FAMILY_PY.get(fn)
    expected_js = ir_kernel.ALGORITHM_FAMILY_JS.get(fn)
    if transcendental["algorithm_py"] != expected_py:
        _fail(
            f"transcendental.algorithm_py {transcendental['algorithm_py']!r} does not match "
            f"the Python reference kernel's own declared family {expected_py!r} for {fn!r}"
        )
    if transcendental["algorithm_js"] != expected_js:
        _fail(
            f"transcendental.algorithm_js {transcendental['algorithm_js']!r} does not match "
            f"the JavaScript twin's own declared family {expected_js!r} for {fn!r}"
        )
    if not isinstance(transcendental["default_terms"], int) or transcendental["default_terms"] <= 0:
        _fail(f"transcendental.default_terms must be a positive integer, got {transcendental['default_terms']!r}")


def _check_domain(value: Fraction, domain: str, name: str) -> None:
    if domain == "Z":
        if value.denominator != 1:
            _fail(f"input {name!r} declares domain Z but {value} is not an integer")
    elif domain == "Z_pos":
        if value.denominator != 1 or value <= 0:
            _fail(f"input {name!r} declares domain Z_pos but {value} is not a positive integer")
    elif domain == "Q":
        pass
    elif domain == "Q_pos":
        if value <= 0:
            _fail(f"input {name!r} declares domain Q_pos but {value} is not positive")
    elif domain == "Q_nonzero":
        if value == 0:
            _fail(f"input {name!r} declares domain Q_nonzero but the given value is zero")
    else:  # pragma: no cover -- validate_ir_shape already rejects any other domain string
        _fail(f"unknown domain: {domain!r}")


def _series_error_bound(fn: str, terms: int) -> Fraction:
    """A conservative, honestly-computed upper bound on the truncation error of this
    reference kernel's own series/iteration at `terms`, given its argument-reduction
    strategy (every reduced argument satisfies |x| <= 1/4 before the series/Newton loop
    runs, per ir_kernel's own reduction constants) -- NOT a measured empirical delta, a
    bound derived from the alternating-series remainder rule (next omitted term's own
    magnitude) applied to the worst case |x| = 1/4, plus a fixed safety factor for the
    recombination steps (squaring/double-angle/sqrt-scaling), which can only shrink a
    already-tiny error further for a contracting map. Always paired with `terms_used` in
    ir_eval.evaluate's return shape (sec.4) so a reader can see both the term count and this
    bound, never one without the other."""
    x = Fraction(1, 4)
    if fn in ("sin", "cos", "pi_const"):
        # alternating series in x**2; next term bound
        k = terms
        bound = (x ** (2 * k + 1)) / ir_kernel._factorial(2 * k + 1) if fn != "pi_const" else (x ** (2 * k + 1))
    elif fn == "exp":
        bound = (x ** (terms + 1)) / ir_kernel._factorial(terms + 1)
    elif fn == "sqrt":
        # Newton-Raphson error roughly squares each step from an initial error <= 1;
        # 2**-(2**terms) is astronomically small for terms >= 5 -- reported honestly, not
        # rounded up to a rounder-looking number.
        bound = Fraction(1, 2 ** min(2 ** terms, 4096))
    elif fn == "log":
        bound = (x ** (2 * terms + 1)) / (2 * terms + 1)
    else:  # pragma: no cover
        bound = Fraction(1, 10 ** 20)
    # Safety margin: never claim a bound tighter than the kernel's own internal rounding
    # precision could actually resolve.
    floor = Fraction(1, 10 ** (ir_kernel.WORKING_PRECISION_DIGITS - 5))
    return max(bound, floor)


def load_ir(path: str | Path) -> dict[str, Any]:
    """Load and structurally validate one IR sidecar file. Raises IRValidationError (never
    returns a partially-trusted document) on any shape mismatch."""
    with open(path, "r", encoding="utf-8") as f:
        ir = json.load(f)
    validate_ir_shape(ir)
    return ir


def evaluate(ir: Mapping[str, Any], inputs: Mapping[str, str]) -> dict[str, Any]:
    """Evaluate an already-loaded, already-validated IR document at the given inputs.
    `inputs` values MUST be strings -- a float/int is rejected here, never coerced, so no
    floating value ever crosses into the evaluator (sec.4/sec.8)."""
    validate_ir_shape(ir)

    input_vars = [v for v in ir["variables"] if v["role"] == "input"]

    env: dict[str, Fraction] = {}
    for v in input_vars:
        name = v["name"]
        if name not in inputs:
            _fail(f"missing required input {name!r}")
        raw = inputs[name]
        if not isinstance(raw, str):
            _fail(f"input {name!r} must be an exact-rational string (e.g. \"3\", \"22/7\"), "
                  f"got {type(raw).__name__} -- a JSON/Python number is never accepted")
        value = ir_kernel.parse_rational(raw)
        _check_domain(value, v["domain"], name)
        env[name] = value

    transcendental = ir.get("transcendental")
    terms = int(transcendental["default_terms"]) if transcendental else ir_kernel.DEFAULT_TERMS

    result = ir_kernel.eval_node(ir["rhs"], env, terms)

    if transcendental is not None:
        terms_used: int | None = terms
        error_bound: Fraction | None = _series_error_bound(transcendental["fn"], terms)
    else:
        terms_used = None
        error_bound = None

    return {
        "value": str(result),
        "terms_used": terms_used,
        "error_bound": str(error_bound) if error_bound is not None else None,
    }


def evaluate_file(path: str | Path, inputs: Mapping[str, str]) -> dict[str, Any]:
    """Convenience wrapper: load, validate, and evaluate one IR sidecar file in one call."""
    return evaluate(load_ir(path), inputs)


def _main(argv: list[str] | None = None) -> int:
    """CLI convenience for manual spot-checks: `python3 -m scripts.executable.ir_eval
    <ir.json> '{"n": "3"}'`. Not used by any other stream -- toledo_eval/crosscheck_runner
    call `evaluate`/`evaluate_file` in-process."""
    import sys

    args = sys.argv[1:] if argv is None else argv
    if len(args) != 2:
        print("usage: ir_eval.py <ir.json> '<json object of string inputs>'", file=sys.stderr)
        return 2
    ir_path, inputs_json = args
    try:
        inputs = json.loads(inputs_json)
        result = evaluate_file(ir_path, inputs)
    except (IRValidationError, ir_kernel.IRKernelError, json.JSONDecodeError) as exc:
        print(json.dumps({"evaluable": False, "reason": str(exc)}), file=sys.stderr)
        return 1
    print(json.dumps({"evaluable": True, **result}, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":  # pragma: no cover
    import sys
    sys.exit(_main())
